Module 2 Project exercise

gmoraita opened the pull request:
- Using the Trello API
Able to:
• Fetch all to-do items (cards) for the specified board
• Create a new card on the board's 'To Do' list
• Move a card from 'To Do' to 'Done' (and 'Doing') - code is agnostic to status and can be more than 3
• Using classes
• Clean code
• Added Description and due date (using date picker)
• Added "Loading" overlay

gmoraita-dev committed:
Initial commit for Module 2 Project

gmoraita-dev committed:
Clean up to some code

gmoraita-dev committed:
Added some commentry

gmoraita-dev committed:
Some more commentry

gmoraita-dev committed:
Minor bugfixes

gmoraita-dev committed:
Minor code simplifications and some UI enhancements

Jonesey13 commented on todo_app/README.md:
```diff
@@ -0,0 +1,8 @@
+Please add a "creds.config" file in the data folder with the following format:
```
This is great (certainly saved me some time reviewing!), but I'd suggest adding this to the existing `README.md` at the project root (or at the very least add a link to here)

I'd also recommend explaining that a developer will need to setup a Trello account and how to find/create their key/token.

gmoraita replied:
Done - moved to root README and updated the instructions as suggested

Jonesey13 commented on todo_app/app.py:
```diff
@@ -1,34 +1,38 @@
-from todo_app.flask_config import Config
-from todo_app.data.session_items import *
+from .data.session_items import add_item, update_item, delete_item, get_items, get_statuses, get_todomapper, TODO_API
+from dateutil import parser
```
`python-dateutil` is not listed in `pyproject.toml` and therefore will not run on a developer's machine after following the README.md instructions.

gmoraita replied:
Done - committed pyproject.toml which I had missed

Jonesey13 commented on todo_app/data/session_items.py:
```diff
@@ -1,93 +1,68 @@
+import json
```
Do we need this file anymore? This feels like it's trying to be an interface layer on top of the `ToDoApi` class which we don't really need and any additional functionality added here could be brought into the class itself.

The singleton instance could be moved to either the bottom of the class file or moved to `app.py`

Jonesey13 replied:
Also the file name (`session_items`) is no longer relevant (as we've stopped using the flask session cookie)

gmoraita replied:
Removed session_items

Jonesey13 commented on todo_app/data/todoapi.py:
```diff
@@ -0,0 +1,160 @@
+        return None
+
+
+class ToDoAPI(ABC):
```
I'd argue this is a case of YAGNI, but hey you never know....

Jonesey13 replied:
Typically you'd have this in a separate file (ideally you'd have 1 file per class)

gmoraita replied:
Agree it is YAGNI. Removed

Jonesey13 commented on todo_app/data/todoapi.py:
```diff
@@ -0,0 +1,160 @@
+
+_TRELLO = 'trello'
+
+def todo_api_factory(api_name):
```
This could become a `@classmethod` on the `TrelloAPI` class

gmoraita replied:
Obsolete - removed

Jonesey13 commented on todo_app/data/todoapi.py:
```diff
@@ -0,0 +1,160 @@
+
+        self.boards = self.get_boards_of_user(self.user) 
+        #Currently defaulting to the first board - in future we can have dropdown
+        self.board_id = self.boards[0].id 
```
This might be unexpected (it was for me!). Ideally you could specify the board you want to edit (maybe as a config value)

gmoraita replied:
Agree - moved to config file (now renamed to trello.config)

Jonesey13 commented on todo_app/data/todoapi.py:
```diff
@@ -0,0 +1,160 @@
+from configparser import ConfigParser
+from .boardelements import *
+import os
+import requests
```
`requests` is not listed in `pyproject.toml` and therefore will not run on a developer's machine after following the README.md instructions.

gmoraita replied:
Committed updated pyproject.toml which I had missed previously

Jonesey13 reviewed (changes requested):
Nearly there! The only blocking issue is that the project does not run out of the box (after following the README.md instructions). 

When I add the new dependencies (`python-dateutil`, `requests`) to poetry then everything appears to be working fine.

Rest of my comments are less important but are still worth reviewing if you can.

gmoraita-dev committed:
Implemented PR feedback:
- removed session_items and updated app.py accordingly
- removed unecessary interface
- updated README at root and removed obsolete one
- fixed minor defect with Board class
- updated pyproject.toml with correct dependencies

hugh-emerson commented on pyproject.toml:
```diff
@@ -8,6 +8,85 @@ authors = []
 python = "^3.7"
 flask = "^1.1.2"
 python-dotenv = "^0.14.0"
+astroid = "2.3.3"
```
You do not need all of these here!
pyproject.toml does not need to include nested dependencies, and indeed *shouldn't*. And it looks like a lot of these aren't even nested dependencies, they are just packages that are totally unused by this project (did you get the list by doing pip freeze on your global python?)

You should just list python, flask, python-dotenv, requests, and python-dateutil.

Delete everything else.

FYI you shouldn't need to update this file manually. Whenever you add a new dependency to the project, use `poetry add <package name>`. That will update pyproject.toml and also install the package.

gmoraita replied:
Agreed - fixed

hugh-emerson commented on todo_app/app.py:
```diff
@@ -1,34 +1,45 @@
 def add():
-    add_item(request.form.get('title'))
+    todoapi.add_item(request.form)
     return render_index_response()
```
This should `return redirect(url_for('index'))`

The difference is the server will send a redirect response to the browser which will then make a new request to the index page. The main benefits are: URL displayed in the browser address bar will be correct, refreshing the page won't resubmit the form.

gmoraita replied:
Doing so returns 306. The intention heer is not do /add at the top but rather have single page

hugh-emerson commented on todo_app/data/todomapper.py:
```diff
@@ -0,0 +1,21 @@
+class ToDoMapper():
```
This class seems a little messy. First of all these are all class attributes rather than instance, so there's no need to create instances of the ToDoMapper.

But really it's just a collection of constants. Or you could maybe use an enum.

Also, rather than having all of Trello's field names stored together here, would it be neater to split them up so that `Item`, `BoardStatus` and `Board` know about the field names relevant to themselves?

gmoraita replied:
Agreed - removed completed and moved these to each respective class

hugh-emerson reviewed (approved):
Nice work. I've left some suggestions but the only important one is to tidy up pyproject.toml.

But it works and you can merge this whenever you're ready.

gmoraita-dev committed:
Incorporated deedback from Hugh:
- Removed todomapper and moved elements to respective classes
- Fixed puproject.toml
